### ISSUE TITLE
Mutation scores differ on same file when run with other files

### ISSUE THREAD

**Issue by pwbriggs23** [stryker-mutator/stryker-js#2989]
**Question**
We can run a mutation test on file A and get 50% score. We can then run a test on file A and file B and the mutation score for file A will be DIFFERENT than 50%. 

**Stryker environment**
    "@stryker-mutator/core": "^4.6.0",
    "@stryker-mutator/karma-runner": "^4.6.0",
    "@stryker-mutator/typescript-checker": "^4.6.0",

    "@stryker-mutator/karma-runner": "^4.6.0",

**Additional context**

**Comment by nicojs** [comment]
Hi @pwbriggs23 

This is unexpected. One reason can be that the number of timeouts differ in the 2 runs. Is that the case? Can you verify that?

**Comment by pwbriggs23** [comment]
That is not the case. The number of timeouts is the same in the 2 runs. 
UPDATE: I have dug into this issue more and have found that if I exclude some of the files, my scores display but when I add them back, they become zero. Will dig more into the culprit files and post more findings here.
Thanks for the reply.

**Comment by pwbriggs23** [comment]
Found the unit test causing the issue...just have to figure out why...seems to have to do with mocking a file upload...hoping to post more details soon.

**Comment by nicojs** [comment]
Wow, that's cool that you found the culprit! 

It might be that you're not cleaning up the mock correctly after the test. StrykerJS will reuse test runner processes, so tests run many times in the same process. 

You can troubleshoot this theory by setting [`maxTestRunnerReuse`](https://stryker-mutator.io/docs/stryker-js/configuration/#maxtestrunnerreuse-number) to 1. This will make sure test runner processes don't get reused at all (but will significantly slow down mutation testing).

**Comment by pwbriggs23** [comment]
So let me try to explain my findings...with the maxTestRunnerReuse set to 1, the problem does not occur, which is good, however, like you stated, the tests take forever to finish. This doesn't sound like a real fix so I did more digging. It turns out that I can trace it back to a while loop used only twice in our application and BOTH create the issue. I rewrite the while loop to use a for loop...issue still occurs. I exempt the UpdateOperator mutant test and the issue goes away. So it is pretty clear what the issue is...what is not clear is how do we resolve it?

**Comment by nicojs** [comment]
Wow, really going all out, I love it! Yeah, using maxTestRunnerReuse of 0 is a shotgun approach. Sure it fixes most problems, but the cure is almost worse then the decease.

A bare bones project is ideal for us. I'm very curious to know what's wrong. Good luck

**Comment by pwbriggs23** [comment]
Ok I have made an example...how do I get these files to you?

**Comment by nicojs** [comment]
You can either upload the zip here, create a public git repository and push it to there, or send me a PM on our stryker slack server : https://join.slack.com/t/stryker-mutator/shared_invite/enQtOTUyMTYyNTg1NDQ0LTU4ODNmZDlmN2I3MmEyMTVhYjZlYmJkOThlNTY3NTM1M2QxYmM5YTM3ODQxYmJjY2YyYzllM2RkMmM1NjNjZjM

**Comment by pwbriggs23** [comment]
[StrykerDemo.zip](https://github.com/stryker-mutator/stryker-js/files/6785392/StrykerDemo.zip)

**Comment by pwbriggs23** [comment]
Ok so the above zip has the 2 services I created with one function each and both have a spec file with a unit test. I also included the test.ts file and the stryker config file as well as 2 result reports...one report shows each service with a mutation score. I then changed the name of the service which adjusted the order of the services and the mutation scores change drastically including wiping out of the second service. Let me know if you need any more files or explanation of anything. Thank you for your time in this matter. Really appreciated.

**Comment by nicojs** [comment]
Thanks! I'm trying it out, but I'm missing a lot of project configuration files:

* package.json and package-lock.json
* tsconfig file(s)
* angular.json file
* karma.conf.js file(s)

Could you also provide those? 

Tip: Maybe it is best to start from scratch with `ng new my-demo` and copy paste files in as needed. Be sure `npx ng test` as well as `npx stryker run` work without error. When you're finished, delete the `node_modules` directory and zip the entire project, that way you are sure I will be able to run it.

**Comment by pwbriggs23** [comment]
[my-demo.zip](https://github.com/stryker-mutator/stryker-js/files/6792242/my-demo.zip)

**Comment by pwbriggs23** [comment]
Any update on this issue?

**Comment by nicojs** [comment]
I've tried to reproduce it but still have some issues:

1. tsconfig references a tsconfig that is located 1 directory up. Can you provide that config? 
   ```diff
   /* To learn more about this file see: https://angular.io/config/tsconfig. */
   {
   +    "extends": "../tsconfig.json",
    "compilerOptions": {
      "outDir": "../out-tsc/e2e",
      "module": "commonjs",
      "target": "es2018",
      "types": [
        "jasmine",
        "node"
      ]
    }
   }
   ```
2. There seems to be a non-public dependency to `@dignity-health/ngx-components`, can I simply remove it? There seems to be a reference to it in angular.json:
   ```diff
   "styles": [
   +     "./node_modules/@dignity-health/ngx-components/scss/dignity-health-components.scss",
        "src/assets/scss/index.scss"
      ],
   ```

**Comment by pwbriggs23** [comment]
1. Darn! I do apologize...will send that file ASAP.
2. This can be removed for sure.

So odd, I mean I created a new project with the command line...not sure why it would have some of this stuff. I don't see another tsconfig file anywhere besides the root and the e2e folder.

**Comment by pwbriggs23** [comment]
[tsconfig.zip](https://github.com/stryker-mutator/stryker-js/files/6832345/tsconfig.zip)
Not seeing that tsconfig file nor do I have that reference

**Comment by pwbriggs23** [comment]
Any progress?

**Comment by pwbriggs23** [comment]
I am going to recreate a barebones project and send you another one

**Comment by nicojs** [comment]
Thanks! I was able to reproduce. 🎉 Definitely strange, maybe a bug in mutant placement 🤷‍♂️

It looks like it has nothing to do with angular, so I could create a small mocha project w/o TS and get the same result.

 I don't have time this weekend, but I hope to be able to pick this up next week

**Comment by pwbriggs23** [comment]
Don't want to bug...hahaha...but any update?

**Comment by nicojs** [comment]
I have had an hour to look into this so far. I now know what the cause of this error is. It is a bug in the `@stryker-mutator/karma-runner` plugin.

The update operator will create an infinite loop in this case:

```diff
while (result) {
-  this.x++;
+  this.x--;
  if (this.x == 10) {
    result = false;
  }
}
```

The infinite loop makes it so that the browser _is no longer responding to ping requests from karma_. This is because javascript is single-threaded, and events being scheduled on the JS event loop will no longer be handled. Eventually, karma reports the error: `"Disconnected reconnect failed before timeout of 2000ms (ping timeout)"` (which I couldn't see in the report, a bug in itself). 

The disconnected timeout in karma is being handled as a "RuntimeError" by the karma-runner plugin. This in itself isn't a bad thing. 

However, the browser is still in its disconnected state, plowing away at the infinite loop. The browser isn't closed or restarted. The karma instance has no browser to run any next mutants in. Each subsequent run is reported as completed without failing tests (indeed, without any tests 🤦‍♂️). Because no tests fail, Stryker will report each of those mutants as survived.

The best solution would be to detect a DISCONNECT in the karma-runner plugin and restart the browser. I currently don't see an API for that in karma, so I've asked for it here: https://github.com/karma-runner/karma/issues/3691

We could fix it in Stryker by adding a check that there is at least 1 test result. If we detect that no tests ran, we can choose to restart the test runner process. Expensive, but the only way I see to recover from this (without a proper karma API).

**Comment by nicojs** [comment]
I've been thinking about this some more and came up with #3023. @pwbriggs23 do you think such a solution would solve this issue?

**Comment by nicojs** [comment]
Hmm. I've thought about it some more and discovered that #3023 will only ever be a best-effort approach. We still need a good solution for this besides that. I will do some more investigation.

**Comment by nicojs** [comment]
Ok, I've figured out a way to implement a browser restart: https://github.com/karma-runner/karma/issues/3691#issuecomment-887751027

I'll implement it in the karma runner whenever a DISCONNECT happens, instead of the fix discussed here https://github.com/stryker-mutator/stryker-js/issues/2989#issuecomment-885461482

**Comment by nicojs** [comment]
We released `v5.2.3` with the fix. Can you confirm that solves your issue?

I'm also busy with implementing #3023 which will land in v5.3. With that, the performance of infinite loop creating mutants will increase dramatically.

**Comment by pwbriggs23** [comment]
It seems that this DOES resolve the issue. I am currently running the entire branch so we will really see if it is fixed. Will update here. Thank you so much for your quick response and your fix.

**Comment by nicojs** [comment]
And? Did it work?

The break-infinite-loops feature also just released, in v5.3. please try it out and let us know if it worked.

**Comment by pwbriggs23** [comment]
Yes...the changes seem to have worked...the mutation test ran on the entire repo and we have a fully scored result. Report looks nice too! Thanks so much for fixing this!
